calyx/src/ir/analysis/graph.rs

```diff
@@ -0,0 +1,159 @@
+        let mut nodes = HashMap::new();
+
+        let mut insert_asgn = |asgn: &Assignment| {
+            let src_key = asgn.src.borrow().key();
+            let dst_key = asgn.dst.borrow().key();
+            nodes
+                .entry(src_key.clone())
+                .or_insert(graph.add_node(Rc::clone(&asgn.src)));
+            nodes
+                .entry(dst_key.clone())
+                .or_insert(graph.add_node(Rc::clone(&asgn.dst)));
+            let src_node = nodes[&src_key];
+            let dst_node = nodes[&dst_key];
+            graph.add_edge(src_node, dst_node, ());
+
+            // add edges for ports
+            if let Some(guard) = &asgn.guard {
+                for port in guard.all_ports() {
+                    let guard_key = port.borrow().key();
+                    nodes
+                        .entry(guard_key.clone())
+                        .or_insert(graph.add_node(Rc::clone(&port)));
+                    graph.add_edge(nodes[&guard_key], dst_node, ());
+                }
+            };
+        };
+
+        for asgn in &component.continuous_assignments {
+            insert_asgn(asgn);
+        }
+
+        for group in &component.groups {
+            for asgn in &group.borrow().assignments {
+                insert_asgn(asgn);
+            }
+        }
+
+        Analysis { nodes, graph }
+    }
+
+    /// Returns an iterator over all the reads from a port. Returns an empty iterator
+    /// if this is an Input port.
+    pub fn reads_from(
+        &self,
+        port: &Port,
+    ) -> impl Iterator<Item = RRC<Port>> + '_ {
+        let idx = self.nodes[&port.key()];
+        match port.direction {
+            Direction::Input => unimplemented!(),
+            Direction::Output | Direction::Inout => {
+                self.graph.edges_directed(idx, Outgoing).map(move |edge| {
+                    let node_idx =
+                        self.graph.edge_endpoints(edge.id()).unwrap().1;
+                    Rc::clone(&self.graph[node_idx])
+                })
+            }
+        }
+    }
+
+    /// Returns an iterator over all the writes to this port. Returns an empty iterator
+    /// if this is an Output port.
+    pub fn writes_to(
+        &self,
+        port: &Port,
+    ) -> impl Iterator<Item = RRC<Port>> + '_ {
+        let idx = self.nodes[&port.key()];
+        match port.direction {
+            Direction::Input | Direction::Inout => {
+                self.graph.edges_directed(idx, Incoming).map(move |edge| {
+                    let node_idx =
+                        self.graph.edge_endpoints(edge.id()).unwrap().0;
+                    Rc::clone(&self.graph[node_idx])
+                })
+            }
+            Direction::Output => unimplemented!(),
+        }
+    }
+
+    /// Restricts the graph of the analysis using a filter
+    /// function that specifies which edges to keep.
+    pub fn edge_induced_subgraph<F>(self, mut f: F) -> Self
```

Comment:
This comment doesn't make sense w.r.t to the function name. What sub-graph is this "inducing"?

Reply:
It is just a subgraph that is created (induced) using a subset of edges. https://mathworld.wolfram.com/Edge-InducedSubgraph.html

Reply:
This makes sense only if you think of assignments as edges. The default representation says nothing about assignments being edges. The comment needs to explicate that.